benches/benches/bevy_ecs/archetype_updates.rs

```diff
@@ -0,0 +1,61 @@
+use bevy_ecs::{
+    component::Component,
+    schedule::{Stage, SystemStage},
+    world::World,
+};
+use criterion::{criterion_group, criterion_main, Criterion};
+
+criterion_group!(benches, no_archetypes, added_archetypes);
+criterion_main!(benches);
+
+#[derive(Component)]
+struct A(f32);
+#[derive(Component)]
+struct B(f32);
+#[derive(Component)]
+struct C(f32);
+#[derive(Component)]
+struct D(f32);
+#[derive(Component)]
+struct E(f32);
+
+const SYSTEMS: usize = 50;
+
+fn setup() -> (World, SystemStage) {
+    let mut world = World::new();
+    fn empty() {}
+    let mut stage = SystemStage::parallel();
+    for _ in 0..SYSTEMS {
+        stage.add_system(empty);
+    }
+    stage.run(&mut world);
+    (world, stage)
+}
+
+fn no_archetypes(criterion: &mut Criterion) {
+    let (mut world, mut stage) = setup();
+    criterion.bench_function("no_archetypes", |bencher| {
+        bencher.iter(|| {
+            stage.run(&mut world);
+        });
+    });
+}
+
+fn added_archetypes(criterion: &mut Criterion) {
```

Comment:
I would like to see this benchmark scaled across several orders of magnitude, probably up to about 10k archetypes.

It'll make any change clearer, and it's valuable to see how performance varies with archetype number regardless.

Ideally we could keep the number of distinct component types the same, and just vary the number of actual archetypes instead.

Reply:
I'm not sure how to add large numbers of archetypes easily - if there's a good way to do that then that would be desirable, yes.

Reply:
can we generate them with a macro and const generics? might get painfully slow to compile.

Reply:
Generated them with const generics and bitwise ops for up to 64K archetypes. Need to go run this new benchmark on the chonky computer and see. :)